### ISSUE TITLE
iFrame Title is not announced by screenreaders as expected

### ISSUE THREAD

**Issue by aeroplanejane** [bbc/simorgh#4891]
Solution here - https://github.com/bbc/simorgh/issues/4891#issuecomment-579837221

Decided against implementing as a new player is incoming and the proposed changes have potential knock-on consequences for other screen readers that we can't predict. Testing resource would be high for this.

**Describe the bug**
We implemented a solution that ensures the iframe title is announced to users. 
This isn't happening - when tabbing to the iframe using talkback on Android no title is announced. 

The issue was: https://github.com/bbc/simorgh/issues/3961

The current Android Talkback announcement order after the page heading is:

1. (iframe appears to be targeted) _nothing_
2. (iframe appears to be targeted) "Graphic"
3. (button appears to be targeted) "Play am, [TITLE] - button - Play am, [TITLE]"

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to a live MAP page (i.e. https://www.bbc.com/pidgin/tori-51268843)
2. Enable Talkback on Android.
3. Swipe to the player
4. Note that the iframe title isn't announced 

[Example video of behaviour as described in issue](https://www.youtube.com/watch?v=p-g3MF6RBio&feature=youtu.be)

**Expected behavior**
Desired behaviour is described in this design https://zpl.io/bzLmZ33

When the user navigates to first iframe for the player 
then "Media Player" is to be announced for the first iframe title
When the user navigates to the second iframe of the player 
then "[headline]" is announced as the second iframe title.
 
This may not occur across all AT testing tools. 

**Smartphone (please complete the following information):**

- Device: Samsung GT-900F (S5)
- OS: Android 6.0.1
- Browser Samsung Browser
- Version 10.2

Using Android Talkback (important - other ATs don't have this problem)

**Testing notes**

- [ ] This bug fix is expected to need manual testing.

**Additional context**

**Comment by LilyL0u** [comment]
On first swiping to the player, the talkback screenreader says nothing, on the next swipe, the focus is still on the frame (second iframe) and it reads out 'graphic'. This happens on audio AND video players.

On Voiceover, nothing is read out for either frame.

**Comment by RichardPK** [comment]
[Example video of behaviour](https://youtu.be/p-g3MF6RBio)

**Comment by RichardPK** [comment]
@aeroplanejane I've updated the issue description with some more details. I've included a reference video for the bug on Android Talkback. It would be good just to confirm that pts 10, 11 & 12 in https://zpl.io/bzLmZ33 describe the expected behaviour that we're shooting for?

On iOS there is also a mismatch with the behaviour described in Zeplin. Demo video:  https://youtu.be/c6s7lraOxvc. It feels quite good but it entirely skips selecting the iFrame and gets all of the video information from the button element.

**Comment by aeroplanejane** [comment]
Behaviour on IOS is good - we'd rather ignore the iframe entirely. On talkback, we can't seem to ignore it, hence the solution above (described in the design) was implemented. Hope this makes sense?

**Comment by RichardPK** [comment]
100%, nice have updated issue

**Comment by aeroplanejane** [comment]
@greenc05 this seems to work as expected on NVDA - is a fix possible for talkback - do we know?

**Comment by greenc05** [comment]
Ok so... As @aeroplanejane says we'd rather ignore the iframe entirely, so if that's happening then awesome! Though that's not possible to achieve across all the screen readers we support. So we defined a desired screen reader UX based on that fact (for screen readers is acknowledging iframes) - which is to announce a title for each iframe so they have context rather than it just saying 'frame' or the like. Though it seems that it was perhaps miss communicated that we don't want to make iframes do this if said screen reader isn't acknowledging the iframe, this is good! 

From my testing notes I have "Note, only NVDA and VoiceOver Mac acknowledge the actual iframe element, the other 3 supported screen readers do not and they will just go straight to the play button. Though note that Talkback has an empty swipe for each iframe - Not sure why..." I believe the empty swipe with Talkback is what we should be investigating here, why is this happening etc? Can we stop it?

**Comment by RichardPK** [comment]
@greenc05 @aeroplanejane tl;dr - it is possible to create the expected behaviour as in this issue within TalkBack. A change is required to the psammead StyledIframe component (we can change ourselves), as well as the SMP iframe (we cannot change ourselves). I don't think it's possible to replicate the current iOS screen-reader behaviour of skipping over the iframes entirely within TalkBack.

Example of the behaviour that can be achieved, dependant on SMP agreeing to make the changes - https://youtu.be/FySVfimMZ7I

-----

So - Talkback 'announcing' an empty frame as well as "graphic" before the button - buckle up.

1. The first announcement - the empty frame. This is the StyledIframe element (within Psammead - `packages/components/psammead-media-player/src/Canonical/index.jsx`). Adding any `aria-label` property to this iFrame will cause TalkBack to read out the element's `title` (note: not actually the value you assign to `aria-label`). 

2. The second announcement - "graphic". This is within second iframe element, this one handled by SMP - `id="smphtml5iframemediaPlayer"`.
a. The "graphic" announcement appears connected to the svg element within this iframe, which contains a lot of `g` elements such as `play`, `play-iplayer`, `subtitles` etc. If we were to apply an `aria-hidden="true"` property to this svg element, while "graphic" will no longer be announced, the screen reader will still 'select' the iframe. It will then either announce nothing, or the iFrame `title` if an `aria-label` property is added to the iFrame (same behaviour as in 1.).  
b. Passing any `aria-label` value into this iFrame without modifying the svg `aria-hidden` property will cause TalkBack to read out "graphic", followed by the iframe `title` property.

Any changes to this SMP iframe are outwith our control.

I have tested making these changes on the iPhone's default screen reader within Safari and they do not alter the existing behaviour (i.e. it goes straight to the button, not selecting either iframe).

Three notes:

Note 1. I have been making these changes within the phone's Chrome / Safari's inspectors, not the code base itself.

Note 2. Assigning the property `aria-hidden="true"` to either of these iFrames makes the iFrame itself as well as all children of the iframe (i.e. the the play button) invisible to TalkBack, so this isn't a solution.

Note 3. If using the localhost pidgin article `pidgin/23248703` - the video title is Japanese and so sometimes won't be announced for language pack reasons, rather than a bug within the aria code.

**Comment by greenc05** [comment]
@RichardPK Good investigation work! I'd be tempted to add the aria-label to the iframe at our end, though that would need to contain the same text as the iframe title, just incase some screen readers acknowledged that, though it's a bit of a hack so I'm slightly hesitant. I say lets leave it as is for now, as there is a new player on the way, all agree?

**Comment by RichardPK** [comment]
OK cool, I'll close for now and reference my comment w/solution in the issue description if ever this issue is reopened.

**Comment by RichardPK** [comment]
Closing issue until we have resource to test the impact of the aria-label change on the MAP iframe.
